File: downloader/downloaders/youtube.py
from typing import List
from pytube import YouTube
from concurrent.futures import ThreadPoolExecutor
from downloader.utils import create_save_directory
from downloader.status_manager import StatusManager


class YouTubeDownloader:
    def __init__(self, save_dir: str, manager: StatusManager, max_workers: int = 5):
        self.save_dir = save_dir
        self.manager = manager
        self.max_workers = max_workers

    def download_video(self, url: str) -> str:
        try:
            yt = YouTube(url)
            stream = yt.streams.get_highest_resolution()
            dir_path = create_save_directory(self.save_dir)
            file_path = stream.download(output_path=dir_path)
            print(f"Downloaded {yt.title} successfully to {file_path}")
            return yt.title
        except Exception as e:
            print(f"Failed to download {url}. Error: {e}")
            return None

    def download_videos(self, urls: List[str]) -> None:
        create_save_directory(self.save_dir)
        with ThreadPoolExecutor(max_workers=self.max_workers) as executor:
            for youtube_url in urls:
                executor.submit(
                    self._download_and_update_status, youtube_url)
        print("All YouTube videos downloaded successfully.")

    def _download_and_update_status(self, url: str) -> None:
        video_title = self.download_video(url)
        if video_title:
            self.manager.update_status(url, video_title, 'Completed')
        else:
            self.manager.update_status(url, video_title, 'Failed')

File: src/main/downloader/downloaders/youtube.ts
import * as path from 'path'
import * as fs from 'fs'
import ytdl from 'ytdl-core'
import { createSaveDirectory } from '../utils'
import { StatusManager } from '../managers/status_manager'

// Define the type for the URL list
type UrlList = string[]

// YouTubeDownloader class
class YouTubeDownloader {
  saveDir: string
  manager: StatusManager
  maxWorkers: number

  constructor(saveDir: string, manager: StatusManager, maxWorkers: number = 5) {
    this.saveDir = saveDir
    this.manager = manager
    this.maxWorkers = maxWorkers
  }

  async downloadVideo(url: string): Promise<string | null> {
    try {
      this.manager.updateStatus(url, 'Downloading')
      const info = await ytdl.getInfo(url)
      const videoTitle = info.videoDetails.title
      const dirPath = createSaveDirectory(this.saveDir)
      const filePath = path.join(dirPath, `${videoTitle}.mp4`)

      const videoStream = ytdl(url, { quality: 'highest' })
      const fileStream = fs.createWriteStream(filePath)

      videoStream.pipe(fileStream)

      return new Promise((resolve, reject) => {
        fileStream.on('finish', () => {
          console.log(`Downloaded ${videoTitle} successfully to ${filePath}`)
          resolve(videoTitle)
        })

        fileStream.on('error', (error) => {
          console.error(`Failed to download ${url}. Error: ${error}`)
          reject(null)
        })
      })
    } catch (error) {
      console.error(`Failed to download ${url}. Error: ${error}`)
      return null
    }
  }

  async downloadVideos(urls: UrlList): Promise<void> {
    createSaveDirectory(this.saveDir)
    const promises = urls.map((url) => this.downloadAndUpdateStatus(url))
    await Promise.all(promises)
    console.log('All YouTube videos downloaded successfully.')
  }

  private async downloadAndUpdateStatus(url: string): Promise<void> {
    const videoTitle = await this.downloadVideo(url)
    if (videoTitle) {
      this.manager.updateStatus(url, 'Completed')
    } else {
      this.manager.updateStatus(url, 'Failed')
    }
  }
}

export { YouTubeDownloader }
